Port WebExtension Icons to C++

ItsJamie9494 opened the pull request:
#### 62bde3cd7c217887a36e2b8f10fe372684b7e6cf
<pre>
Port WebExtension Icons to C++
<a href="https://webkit.org/b/282500">https://webkit.org/b/282500</a>

Reviewed by Timothy Hatcher.

Replace CocoaImage with WebCore::Icon, as WebCore::Icon wraps CocoaImage already.
Port most icon utility functions to C++, and for those that cannot be ported,
update them to support a cross-platform API involving WebCore::Icon and JSON::Value.

* Source/WTF/wtf/JSONValues.h:
* Source/WebKit/Platform/cocoa/CocoaHelpers.h:
* Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
(WebKit::availableScreenScales): Deleted.
(WebKit::largestDisplayScale): Deleted.
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.cpp:
(WebKit::largestDisplayScale):
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.h:
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.mm:
(WebKit::availableScreenScales):
* Source/WebKit/UIProcess/API/Cocoa/WKWebExtension.mm:
(-[WKWebExtension iconForSize:]):
(-[WKWebExtension actionIconForSize:]):
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIActionCocoa.mm:
(WebKit::WebExtensionContext::actionSetIcon):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionActionCocoa.mm:
(WebKit::WebExtensionAction::clearCustomizations):
(WebKit::WebExtensionAction::icon):
(WebKit::WebExtensionAction::setIcons):
(WebKit::WebExtensionAction::setIconVariants):
(WebKit::WebExtensionAction::clearIconCache):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::imageForPath):
(WebKit::WebExtension::bestImageInIconsDictionary):
(WebKit::WebExtension::bestImageForIconVariants):
(WebKit::WebExtension::icon): Deleted.
(WebKit::WebExtension::actionIcon): Deleted.
(WebKit::WebExtension::bestSizeInIconsDictionary): Deleted.
(WebKit::WebExtension::pathForBestImageInIconsDictionary): Deleted.
(WebKit::WebExtension::bestImageForIconsDictionaryManifestKey): Deleted.
(WebKit::toColorSchemes): Deleted.
(WebKit::WebExtension::iconsDictionaryForBestIconVariant): Deleted.
(WebKit::WebExtension::bestImageForIconVariantsManifestKey): Deleted.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCommandCocoa.mm:
(WebKit::WebExtensionCommand::platformMenuItem const):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMenuItemCocoa.mm:
(WebKit::WebExtensionMenuItem::WebExtensionMenuItem):
(WebKit::WebExtensionMenuItem::update):
(WebKit::WebExtensionMenuItem::icon const):
(WebKit::WebExtensionMenuItem::clearIconCache const):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
(WebKit::getDefaultIconsDictFromExtension):
(WebKit::WebExtensionSidebar::icon):
(WebKit::WebExtensionSidebar::setIconsDictionary):
(): Deleted.
* Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
(WebKit::WebExtension::icon):
(WebKit::WebExtension::actionIcon):
(WebKit::WebExtension::bestSizeInIconsDictionary):
(WebKit::WebExtension::pathForBestImageInIconsDictionary):
(WebKit::WebExtension::bestImageForIconsDictionaryManifestKey):
(WebKit::toColorSchemes):
(WebKit::WebExtension::iconsDictionaryForBestIconVariant):
(WebKit::WebExtension::bestImageForIconVariantsManifestKey):
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
(WebKit::WebExtension::imageForPath):
* Source/WebKit/UIProcess/Extensions/WebExtensionAction.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionMenuItem.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionSidebar.h:

Canonical link: <a href="https://commits.webkit.org/286237@main">https://commits.webkit.org/286237@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ac00f860883605d82274269b34a7ed3a6c7b3d70

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75287 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/54726 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28127 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/79749 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/26543 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/63866 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/2511 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/59089 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/17327 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78354 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49271 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64681 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/39465 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46667 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/22175 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/24871 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/68434 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/67684 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/22514 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81233 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/74550 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/2617 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/1631 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/67337 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/2768 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64678 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/66629 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/10577 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/8734 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/96818 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11624 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/2578 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/5393 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21147 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/2603 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/3533 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/2612 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionActionCocoa.mm:
```diff
@@ -765,31 +765,33 @@ - (void)_otherPopoverWillShow:(NSNotification *)notification
         return;
 
-    m_customIcons = icons.count ? icons : nil;
+    m_customIcons = icons ?: nullptr;
```
I think this should this check `size()`.
```suggestion
    m_customIcons = icons.size() ? icons : nullptr;
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -673,123 +614,71 @@ static String processFileAndExtractZipArchive(const String& path)
 #if PLATFORM(IOS_FAMILY)
     auto *scalePaths = [NSMutableDictionary dictionary];
 #endif
+    RefPtr iconsObject = icons->asObject();
```
```suggestion

    RefPtr iconsObject = icons->asObject();
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -673,123 +614,71 @@ static String processFileAndExtractZipArchive(const String& path)
 
 #if PLATFORM(IOS_FAMILY)
-        scalePaths[scale] = iconPath;
+        scalePaths[[NSNumber numberWithDouble:scale]] = iconPath;
```
Can use the box shorthand for this.
```suggestion
        scalePaths[@(scale)] = iconPath;
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -825,137 +714,38 @@ static String processFileAndExtractZipArchive(const String& path)
 
     // If the light and dark icons dictionaries are the same, or if either is nil, return the available image directly.
-    if (!lightIconsDictionary || !darkIconsDictionary || [lightIconsDictionary isEqualToDictionary:darkIconsDictionary])
+    if (!lightIconsDictionary || !darkIconsDictionary || (lightIconsDictionary == darkIconsDictionary))
```
```suggestion
    if (!lightIconsDictionary || !darkIconsDictionary || lightIconsDictionary == darkIconsDictionary)
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMenuItemCocoa.mm:
```diff
@@ -189,10 +189,13 @@ + (BOOL)usesUserKeyEquivalents
+        if (parsedIcons)
+            m_iconVariants = parsedIcons->asArray();
+        else
+            m_iconVariants = nullptr;
```
```suggestion
        m_iconVariants = parsedIcons ? parsedIcons->asArray() : nullptr;
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMenuItemCocoa.mm:
```diff
@@ -115,10 +115,10 @@ + (BOOL)usesUserKeyEquivalents
+        m_icons = parsedIcons;
 #if ENABLE(WK_WEB_EXTENSIONS_ICON_VARIANTS)
-        m_iconVariants = dynamic_objc_cast<NSArray>(parsedIcons);
+        m_iconVariants = parsedIcons->asArray();
```
```suggestion
        m_iconVariants = parsedIcons ? parsedIcons->asArray() : nullptr;
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMenuItemCocoa.mm:
```diff
@@ -115,10 +115,10 @@ + (BOOL)usesUserKeyEquivalents
-        id parsedIcons = parseJSON(parameters.iconsJSON, JSONOptions::FragmentsAllowed);
-        m_icons = dynamic_objc_cast<NSDictionary>(parsedIcons);
+        RefPtr parsedIcons = JSON::Value::parseJSON(parameters.iconsJSON);
+        m_icons = parsedIcons;
```
```suggestion
        m_icons = parsedIcons ? parsedIcons->asObject() : nullptr;
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMenuItemCocoa.mm:
```diff
@@ -189,10 +189,13 @@ + (BOOL)usesUserKeyEquivalents
-        id parsedIcons = parseJSON(parameters.iconsJSON, JSONOptions::FragmentsAllowed);
-        m_icons = dynamic_objc_cast<NSDictionary>(parsedIcons);
+        RefPtr parsedIcons = JSON::Value::parseJSON(parameters.iconsJSON);
+        m_icons = parsedIcons;
```
```suggestion
        m_icons = parsedIcons ? parsedIcons->asObject() : nullptr;
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
```diff
@@ -316,27 +316,27 @@ - (void)otherMouseDown:(NSEvent *)event
     }
 
-    if (m_iconsOverride && [m_iconsOverride.value() isEqualToDictionary:icons])
+    if (m_iconsOverride && (m_iconsOverride.value() == icons))
```
```suggestion
    if (m_iconsOverride && m_iconsOverride.value() == icons)
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -1548,6 +1557,279 @@ void WebExtension::populatePermissionsPropertiesIfNeeded()
+        return nullptr;
+
+    if (m_defaultActionIcon)
+        return m_defaultActionIcon;
```
This should move up, under `populateActionPropertiesIfNeeded()` like before.

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -1548,6 +1557,279 @@ void WebExtension::populatePermissionsPropertiesIfNeeded()
+#endif
+
+    // Check if the ideal size exists, if so return it.
+    String idealSizeString = String::number(idealPixelSize);
```
```suggestion
    auto idealSizeString = String::number(idealPixelSize);
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -1548,6 +1557,279 @@ void WebExtension::populatePermissionsPropertiesIfNeeded()
+    if (iconsObject.getValue(idealSizeString))
+        return idealPixelSize;
+
+    // Sort the remaining keys and find the next largest size.
```
I'd move this comment down to the `std::sort` or change this to be more about making a `Vector` of integer values.

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionAction.h:
```diff
@@ -91,9 +92,9 @@ class WebExtensionAction : public API::ObjectImpl<API::Object::Type::WebExtensio
     void propertiesDidChange();
 
     CocoaImage *icon(CGSize);
```
This should change to match `WebExtension`. It would simplify that implementation a bit.
```suggestion
    RefPtr<WebCore::Icon> icon(WebCore::FloatSize idealSize);
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -1548,6 +1557,279 @@ void WebExtension::populatePermissionsPropertiesIfNeeded()
+    auto cacheKey = idealSize.toJSONString();
+    if (auto cacheResult = cacheLocation.getOptional(cacheKey)) {
+        if (std::holds_alternative<RefPtr<WebCore::Icon>>(*cacheResult))
+            return *std::get_if<RefPtr<WebCore::Icon>>(&*cacheResult);
```
The old code just did this. It will be `nullptr` if the variant is not an icon, and that is fine. That allows cache hits to have null results, to skip extra work each time.
```suggestion
        return *std::get_if<RefPtr<WebCore::Icon>>(&*cacheResult);
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.h:
```diff
@@ -63,6 +64,8 @@ class WebExtension : public API::ObjectImpl<API::Object::Type::WebExtension>, pu
     WTF_MAKE_NONCOPYABLE(WebExtension);
 
 public:
+    using IconVariant = std::variant<RefPtr<WebCore::Icon>, Vector<double>>;
```
`IconVariant` is confusing, since this isn't about extension icon variants but `std::variant`.
```suggestion
    using IconCacheEntry = std::variant<RefPtr<WebCore::Icon>, Vector<double>>;
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -1548,6 +1557,279 @@ void WebExtension::populatePermissionsPropertiesIfNeeded()
+        return nullptr;
+    }
+
+    cacheLocation.set(cacheKey, IconVariant(result));
```
Is `IconVariant()` needed? I'm not sure if `std::variant` can be created with one of the alternatives. But this would be better if it works (applies to all other places that do this):
```suggestion
    cacheLocation.set(cacheKey, result);
```

ItsJamie9494 replied:
When I tried it without specifying IconVariant, it wouldn't allow me to create an std::variant from the type. So as far as I can tell, it is needed.

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -1548,6 +1557,279 @@ void WebExtension::populatePermissionsPropertiesIfNeeded()
+    using ColorScheme = WebExtension::ColorScheme;
+
+    if (!value) {
+        // A nil or invalid value counts as all color schemes.
```
```suggestion
        // A null or invalid value counts as all color schemes.
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -1548,6 +1557,279 @@ void WebExtension::populatePermissionsPropertiesIfNeeded()
+        return nullptr;
+
+    if (variants->length() == 1)
+        return variants->begin()->ptr();
```
Avoid making an `iterator`.
```suggestion
        return variants->get(0)->ptr();
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -1548,6 +1557,279 @@ void WebExtension::populatePermissionsPropertiesIfNeeded()
+    auto cacheKey = idealSize.toJSONString();
+    if (auto cacheResult = cacheLocation.getOptional(cacheKey)) {
+        if (std::holds_alternative<RefPtr<WebCore::Icon>>(*cacheResult))
+            return *std::get_if<RefPtr<WebCore::Icon>>(&*cacheResult);
```
```suggestion
        return *std::get_if<RefPtr<WebCore::Icon>>(&*cacheResult);
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionAction.h:
```diff
@@ -176,12 +177,12 @@ class WebExtensionAction : public API::ObjectImpl<API::Object::Type::WebExtensio
     String m_popupWebViewInspectionName;
 
     RetainPtr<CocoaImage> m_cachedIcon;
```
This should also use `WebCore::Icon` for ease of the impl.

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionMenuItem.h:
```diff
@@ -136,12 +136,12 @@ class WebExtensionMenuItem : public RefCounted<WebExtensionMenuItem>, public Can
     RefPtr<WebExtensionCommand> m_command;
 
     mutable RetainPtr<CocoaImage> m_cachedIcon;
```
Also `WebCore::Icon`.

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionMenuItem.h:
```diff
@@ -136,12 +136,12 @@ class WebExtensionMenuItem : public RefCounted<WebExtensionMenuItem>, public Can
     mutable RetainPtr<CocoaImage> m_cachedIcon;
-    mutable RetainPtr<NSSet> m_cachedIconScales;
+    mutable Vector<double> m_cachedIconScales;
     mutable CGSize m_cachedIconIdealSize { CGSizeZero };
```
`WebCore::FloatSize`

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionAction.h:
```diff
@@ -176,12 +177,12 @@ class WebExtensionAction : public API::ObjectImpl<API::Object::Type::WebExtensio
     RetainPtr<CocoaImage> m_cachedIcon;
-    RetainPtr<NSSet> m_cachedIconScales;
+    Vector<double> m_cachedIconScales;
     CGSize m_cachedIconIdealSize { CGSizeZero };
```
`WebCore::FloatSize`

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionSidebar.h:
```diff
@@ -83,7 +83,7 @@ class WebExtensionSidebar : public API::ObjectImpl<API::Object::Type::WebExtensi
 
     /// `icon()` will return the overridden icon of this sidebar, or the icon of the first parent sidebar in which the icon is set
     RetainPtr<CocoaImage> icon(CGSize);
```
Should also match `WebExtension` using `WebCore::Icon` and `WebCore::FloatSize`.

xeenon reviewed (changes requested):
Looking really good!

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -1548,6 +1557,279 @@ void WebExtension::populatePermissionsPropertiesIfNeeded()
+
+    populateActionPropertiesIfNeeded();
+
+    RefPtr<JSON::Object> actionDictionary;
```
```suggestion
    RefPtr<JSON::Object> actionObject;
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.h:
```diff
@@ -270,18 +273,18 @@ class WebExtension : public API::ObjectImpl<API::Object::Type::WebExtension>, pu
-    size_t bestSizeInIconsDictionary(NSDictionary *, size_t idealPixelSize);
-    NSString *pathForBestImageInIconsDictionary(NSDictionary *, size_t idealPixelSize);
+    size_t bestSizeInIconsDictionary(const JSON::Object&, size_t idealPixelSize);
+    String pathForBestImageInIconsDictionary(const JSON::Object&, size_t idealPixelSize);
```
```suggestion
    size_t bestIconSize(const JSON::Object&, size_t idealPixelSize);
    String pathForBestImage(const JSON::Object&, size_t idealPixelSize);
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.h:
```diff
@@ -270,18 +273,18 @@ class WebExtension : public API::ObjectImpl<API::Object::Type::WebExtension>, pu
-    CocoaImage *bestImageInIconsDictionary(NSDictionary *, CGSize idealSize, const Function<void(Ref<API::Error>)>&);
-    CocoaImage *bestImageForIconsDictionaryManifestKey(NSDictionary *, NSString *manifestKey, CGSize idealSize, RetainPtr<NSMutableDictionary>& cacheLocation, Error, NSString *customLocalizedDescription);
+    RefPtr<WebCore::Icon> bestImageInIconsDictionary(RefPtr<JSON::Value>, WebCore::FloatSize idealSize, const Function<void(Ref<API::Error>)>&);
+    RefPtr<WebCore::Icon> bestImageForIconsDictionaryManifestKey(const JSON::Object&, const String& manifestKey, WebCore::FloatSize idealSize, IconsCache& cacheLocation, Error, const String& customLocalizedDescription);
```
```suggestion
    RefPtr<WebCore::Icon> bestIcon(RefPtr<JSON::Object>, WebCore::FloatSize idealSize, const Function<void(Ref<API::Error>)>&);
    RefPtr<WebCore::Icon> bestIconForManifestKey(const JSON::Object&, const String& manifestKey, WebCore::FloatSize idealSize, IconsCache& cacheLocation, Error, const String& customLocalizedDescription);
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -1548,6 +1557,279 @@ void WebExtension::populatePermissionsPropertiesIfNeeded()
+
+    if (!result) {
+        if (iconObject->size()) {
+            // Record an error if the dictionary had values, meaning the likely failure is the images were missing on disk or bad format.
```
```suggestion
            // Record an error if the object had values, meaning the likely failure is the images were missing on disk or bad format.
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -1548,6 +1557,279 @@ void WebExtension::populatePermissionsPropertiesIfNeeded()
+            // Record an error if the dictionary had values, meaning the likely failure is the images were missing on disk or bad format.
+            recordError(createError(error, customLocalizedDescription));
+        } else if (!iconObject->size() || object.getValue(manifestKey)) {
+            // Record an error if the key had dictionary that was empty, or the key had a value of the wrong type.
```
```suggestion
            // Record an error if the key had object that was empty, or the key had a value of the wrong type.
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -1548,6 +1557,279 @@ void WebExtension::populatePermissionsPropertiesIfNeeded()
+
+    if (!result) {
+        if (iconArray->length()) {
+            // Record an error if the dictionary had values, meaning the likely failure is the images were missing on disk or bad format.
```
```suggestion
            // Record an error if the array had values, meaning the likely failure is the images were missing on disk or bad format.
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -1548,6 +1557,279 @@ void WebExtension::populatePermissionsPropertiesIfNeeded()
+            // Record an error if the dictionary had values, meaning the likely failure is the images were missing on disk or bad format.
+            recordError(createError(error, customLocalizedDescription));
+        } else if (!iconArray->length() || object.getValue(manifestKey)) {
+            // Record an error if the key had dictionary that was empty, or the key had a value of the wrong type.
```
```suggestion
            // Record an error if the key had an array that was empty, or the key had a value of the wrong type.
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.h:
```diff
@@ -270,18 +273,18 @@ class WebExtension : public API::ObjectImpl<API::Object::Type::WebExtension>, pu
-    CocoaImage *bestImageForIconVariantsManifestKey(NSDictionary *, NSString *manifestKey, CGSize idealSize, RetainPtr<NSMutableDictionary>& cacheLocation, Error, NSString *customLocalizedDescription);
+    RefPtr<JSON::Value> iconsDictionaryForBestIconVariant(RefPtr<JSON::Array>, size_t idealPixelSize, ColorScheme);
+    RefPtr<WebCore::Icon> bestImageForIconVariants(RefPtr<JSON::Array>, WebCore::FloatSize idealSize, const Function<void(Ref<API::Error>)>&);
+    RefPtr<WebCore::Icon> bestImageForIconVariantsManifestKey(const JSON::Object&, const String& manifestKey, WebCore::FloatSize idealSize, IconsCache& cacheLocation, Error, const String& customLocalizedDescription);
```
```suggestion
    RefPtr<JSON::Object> bestIconVariantJSONObject(RefPtr<JSON::Array>, size_t idealPixelSize, ColorScheme);
    RefPtr<WebCore::Icon> bestIconVariant(RefPtr<JSON::Array>, WebCore::FloatSize idealSize, const Function<void(Ref<API::Error>)>&);
    RefPtr<WebCore::Icon> bestIconVariantForManifestKey(const JSON::Object&, const String& manifestKey, WebCore::FloatSize idealSize, IconsCache& cacheLocation, Error, const String& customLocalizedDescription);
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.h:
```diff
@@ -270,18 +273,18 @@ class WebExtension : public API::ObjectImpl<API::Object::Type::WebExtension>, pu
 #endif
 
-    CocoaImage *imageForPath(NSString *, RefPtr<API::Error>&, CGSize sizeForResizing = CGSizeZero);
+    RefPtr<WebCore::Icon> imageForPath(NSString *, RefPtr<API::Error>&, WebCore::FloatSize sizeForResizing = { });
```
```suggestion
    RefPtr<WebCore::Icon> iconForPath(NSString *, RefPtr<API::Error>&, WebCore::FloatSize sizeForResizing = { });
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIActionCocoa.mm:
```diff
@@ -125,19 +125,19 @@
-    if (auto *dictionary = dynamic_objc_cast<NSDictionary>(parsedIcons))
-        webExtensionAction->setIcons(dictionary);
+    if (parsedIcons->asObject())
+        webExtensionAction->setIcons(parsedIcons->asObject());
```
```suggestion
    if (RefPtr object = parsedIcons->asObject())
        webExtensionAction->setIcons(object);
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIActionCocoa.mm:
```diff
@@ -125,19 +125,19 @@
     }
 
-    id parsedIcons = parseJSON(iconsJSON, JSONOptions::FragmentsAllowed);
+    RefPtr<JSON::Value> parsedIcons = JSON::Value::parseJSON(iconsJSON);
```
```suggestion
    RefPtr parsedIcons = JSON::Value::parseJSON(iconsJSON);
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionActionCocoa.mm:
```diff
@@ -765,31 +765,33 @@ - (void)_otherPopoverWillShow:(NSNotification *)notification
-    return extensionContext()->extension().actionIcon(idealSize);
+    if (RefPtr icon = extensionContext()->extension().actionIcon(WebCore::FloatSize(idealSize)))
+        return icon;
+    return nullptr;
```
```suggestion
    return extensionContext()->extension().actionIcon(WebCore::FloatSize(idealSize));
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -673,123 +614,68 @@ static String processFileAndExtractZipArchive(const String& path)
                 resultImage = image;
             else
-                [resultImage addRepresentations:image.representations];
+                [resultImage->image().get() addRepresentations:image->image().get().representations];
```
No need for `.get()` when calling with bracket notation.
```suggestion
                [resultImage->image() addRepresentations:image->image().get().representations];
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -825,137 +711,38 @@ static String processFileAndExtractZipArchive(const String& path)
-    auto *darkIconsDictionary = iconsDictionaryForBestIconVariant(variants, idealPointSize, ColorScheme::Dark);
+    auto idealPointSize = idealSize.width() > idealSize.height() ? idealSize.width() : idealSize.height();
+    RefPtr lightIconsDictionary = bestIconVariantJSONObject(variants, idealPointSize, ColorScheme::Light);
+    RefPtr darkIconsDictionary = bestIconVariantJSONObject(variants, idealPointSize, ColorScheme::Dark);
```
```suggestion
    RefPtr lightIconsObject = bestIconVariantJSONObject(variants, idealPointSize, ColorScheme::Light);
    RefPtr darkIconsObject = bestIconVariantJSONObject(variants, idealPointSize, ColorScheme::Dark);
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMenuItemCocoa.mm:
```diff
@@ -298,8 +298,8 @@ + (BOOL)usesUserKeyEquivalents
-    auto idealSize = CGSizeMake(16, 16);
-    auto *image = icon(idealSize);
+    auto idealSize = WebCore::FloatSize(16, 16);
+    auto *image = icon(idealSize) ? icon(idealSize)->image().get() : nil;
```
This should save the result of `icon(idealSize)` and use it instead of calling it again (it can be expensive).

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMenuItemCocoa.mm:
```diff
@@ -342,11 +342,11 @@ + (BOOL)usesUserKeyEquivalents
     }
 
-    return [UIMenu menuWithTitle:processedTitle image:icon(CGSizeMake(20, 20)) identifier:nil options:0 children:submenuItemArray];
+    return [UIMenu menuWithTitle:processedTitle image:icon({ 20, 20 })->image().get() identifier:nil options:0 children:submenuItemArray];
```
Ditto on crashing.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMenuItemCocoa.mm:
```diff
@@ -325,7 +325,7 @@ + (BOOL)usesUserKeyEquivalents
     // iOS does not support sub-menus that are disabled or hidden, so return a normal action in that case.
     if (submenuItems().isEmpty() || !isEnabled() || !isVisible()) {
-        auto *action = [UIAction actionWithTitle:processedTitle image:icon(CGSizeMake(20, 20)) identifier:nil handler:makeBlockPtr([this, protectedThis = Ref { *this }, contextParameters](UIAction *) mutable {
+        auto *action = [UIAction actionWithTitle:processedTitle image:(icon({ 20, 20 })->image().get()) identifier:nil handler:makeBlockPtr([this, protectedThis = Ref { *this }, contextParameters](UIAction *) mutable {
```
I think this is causing a crash on iOS, since you don't null check `icon({ 20, 20 })`.
```suggestion
        UIImage *image;
        if (RefPtr icon = this->icon({ 20, 20 }))
            image = icon->image().get();
        auto *action = [UIAction actionWithTitle:processedTitle image:image identifier:nil handler:makeBlockPtr([this, protectedThis = Ref { *this }, contextParameters](UIAction *) mutable {
```

We should consider a helper function for this, so it is easy to get a `nil` for a `null` `WebCore::Icon`.

```
static inline CocoaImage *toCocoaImage(const RefPtr<WebCore::Icon>& icon)
{
    return icon ? icon->image().get() : nil;
}
```

That needs to return `RetainPtr<CocoaImage>` if you out it in `WebCore/Icon.h`, since that code isn't ObjC ARC.

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -1548,6 +1557,275 @@ void WebExtension::populatePermissionsPropertiesIfNeeded()
+
+String WebExtension::pathForBestImage(const JSON::Object& iconsObject, size_t idealPixelSize)
+{
+    size_t bestSize = bestIconSize(iconsObject, idealPixelSize);
```
```suggestion
    auto bestSize = bestIconSize(iconsObject, idealPixelSize);
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -1548,6 +1557,275 @@ void WebExtension::populatePermissionsPropertiesIfNeeded()
+    size_t fallbackSize = 0;
+
+    // Pick the first variant matching color scheme and/or size.
+    for (Ref<JSON::Value> variant : *variants) {
```
```suggestion
    for (Ref variant : *variants) {
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -1548,6 +1557,275 @@ void WebExtension::populatePermissionsPropertiesIfNeeded()
+            if (currentBestSize > bestSize) {
+                // Found a larger ideal variant.
+                bestSize = currentBestSize;
+                bestVariant = variant->asObject();
```
```suggestion
                bestVariant = variantObject;
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -1548,6 +1557,275 @@ void WebExtension::populatePermissionsPropertiesIfNeeded()
+        if (colorSchemes.contains(idealColorScheme)) {
+            if (currentBestSize >= idealPixelSize) {
+                // Found the best variant, return it.
+                return variant->asObject();
```
```suggestion
                return variantObject;
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -1548,6 +1557,275 @@ void WebExtension::populatePermissionsPropertiesIfNeeded()
+        } else if (!foundIdealFallbackVariant && currentBestSize >= idealPixelSize) {
+            // Found an ideal fallback variant, based only on size.
+            fallbackSize = currentBestSize;
+            fallbackVariant = variant->asObject();
```
```suggestion
            fallbackVariant = variantObject;
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -1548,6 +1557,275 @@ void WebExtension::populatePermissionsPropertiesIfNeeded()
+        } else if (!foundIdealFallbackVariant && currentBestSize > fallbackSize) {
+            // Found a smaller fallback variant.
+            fallbackSize = currentBestSize;
+            fallbackVariant = variant->asObject();
```
```suggestion
            fallbackVariant = variantObject;
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionSidebar.h:
```diff
@@ -131,7 +130,7 @@ class WebExtensionSidebar : public API::ObjectImpl<API::Object::Type::WebExtensi
     void reloadWebView();
 
-    std::optional<RetainPtr<NSDictionary>> m_iconsOverride;
+    std::optional<RefPtr<JSON::Value>> m_iconsOverride;
```
```suggestion
    std::optional<RefPtr<JSON::Object>> m_iconsOverride;
```

xeenon reviewed (changes requested):
Looking good. Need to fix iOS crash in menu code though.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMenuItemCocoa.mm:
```diff
@@ -325,7 +331,7 @@ + (BOOL)usesUserKeyEquivalents
     // iOS does not support sub-menus that are disabled or hidden, so return a normal action in that case.
     if (submenuItems().isEmpty() || !isEnabled() || !isVisible()) {
-        auto *action = [UIAction actionWithTitle:processedTitle image:icon(CGSizeMake(20, 20)) identifier:nil handler:makeBlockPtr([this, protectedThis = Ref { *this }, contextParameters](UIAction *) mutable {
+        auto *action = [UIAction actionWithTitle:processedTitle image:(toCocoaImage(icon({ 20, 20 }))) identifier:nil handler:makeBlockPtr([this, protectedThis = Ref { *this }, contextParameters](UIAction *) mutable {
```
```suggestion
        auto *action = [UIAction actionWithTitle:processedTitle image:toCocoaImage(icon({ 20, 20 })) identifier:nil handler:makeBlockPtr([this, protectedThis = Ref { *this }, contextParameters](UIAction *) mutable {
```

ItsJamie9494 committed:
Port WebExtension Icons to C++
https://webkit.org/b/282500

Reviewed by Timothy Hatcher.

Replace CocoaImage with WebCore::Icon, as WebCore::Icon wraps CocoaImage already.
Port most icon utility functions to C++, and for those that cannot be ported,
update them to support a cross-platform API involving WebCore::Icon and JSON::Value.

* Source/WTF/wtf/JSONValues.h:
* Source/WebKit/Platform/cocoa/CocoaHelpers.h:
* Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
(WebKit::availableScreenScales): Deleted.
(WebKit::largestDisplayScale): Deleted.
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.cpp:
(WebKit::largestDisplayScale):
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.h:
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.mm:
(WebKit::availableScreenScales):
* Source/WebKit/UIProcess/API/Cocoa/WKWebExtension.mm:
(-[WKWebExtension iconForSize:]):
(-[WKWebExtension actionIconForSize:]):
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIActionCocoa.mm:
(WebKit::WebExtensionContext::actionSetIcon):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionActionCocoa.mm:
(WebKit::WebExtensionAction::clearCustomizations):
(WebKit::WebExtensionAction::icon):
(WebKit::WebExtensionAction::setIcons):
(WebKit::WebExtensionAction::setIconVariants):
(WebKit::WebExtensionAction::clearIconCache):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::imageForPath):
(WebKit::WebExtension::bestImageInIconsDictionary):
(WebKit::WebExtension::bestImageForIconVariants):
(WebKit::WebExtension::icon): Deleted.
(WebKit::WebExtension::actionIcon): Deleted.
(WebKit::WebExtension::bestSizeInIconsDictionary): Deleted.
(WebKit::WebExtension::pathForBestImageInIconsDictionary): Deleted.
(WebKit::WebExtension::bestImageForIconsDictionaryManifestKey): Deleted.
(WebKit::toColorSchemes): Deleted.
(WebKit::WebExtension::iconsDictionaryForBestIconVariant): Deleted.
(WebKit::WebExtension::bestImageForIconVariantsManifestKey): Deleted.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCommandCocoa.mm:
(WebKit::WebExtensionCommand::platformMenuItem const):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMenuItemCocoa.mm:
(WebKit::WebExtensionMenuItem::WebExtensionMenuItem):
(WebKit::WebExtensionMenuItem::update):
(WebKit::WebExtensionMenuItem::icon const):
(WebKit::WebExtensionMenuItem::clearIconCache const):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
(WebKit::getDefaultIconsDictFromExtension):
(WebKit::WebExtensionSidebar::icon):
(WebKit::WebExtensionSidebar::setIconsDictionary):
(): Deleted.
* Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
(WebKit::WebExtension::icon):
(WebKit::WebExtension::actionIcon):
(WebKit::WebExtension::bestSizeInIconsDictionary):
(WebKit::WebExtension::pathForBestImageInIconsDictionary):
(WebKit::WebExtension::bestImageForIconsDictionaryManifestKey):
(WebKit::toColorSchemes):
(WebKit::WebExtension::iconsDictionaryForBestIconVariant):
(WebKit::WebExtension::bestImageForIconVariantsManifestKey):
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
(WebKit::WebExtension::imageForPath):
* Source/WebKit/UIProcess/Extensions/WebExtensionAction.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionMenuItem.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionSidebar.h:

Canonical link: https://commits.webkit.org/286237@main
